### PR TITLE
Better rvalues support for parallel_reduce algorithm

### PR DESCRIPTION
**Pull request by kboyarinov**
### Description 
Adding the optimization for reduction of rvalues using `parallel_reduce` and `parallel_deterministic_reduce`.
Based on #171 , #169 and #1299.

This patch is intended to optimize operating with Value class in parallel reduce algorithm (lambda version). Example from #1299

```
// We want to efficiently reduce a number of std::set objects into single std::set containing unique values from each set
std::vector<std::set<int>> sets = ...;

parallel_reduce(blocked_range<size_t>(0, sets.size()),
                         std::set<int>{}, // Identity - empty set
                         [&](const blocked_range<size_t>& range, const std::set<int>& value) { // Second argument is required to be const Value& by the spec
                             std::set<int> result = value; // We are forced to do copy here since value is constant
                             
                             for (size_t i = range.begin(); i < range.end(); ++i) {
                                 result.merge(sets[i]); // set::merge is efficient since it transfers required node from one set to another
                             }
                             return result;
                         },
                         [](const std::set<int>& value1, const std::set<int>& value2) { // Both arguments are required to be const Value&
                             std::set<int> result = value1; // We are forced to do copy
                             result.insert(value2.begin(), value2.end()); // unable to use set::merge here since value2 is constant  
                         }                  
                         );
```

The spec forces the user to use inefficient copying instead of "moving" even if the input range preservation is not important for the user. "Body" version of parallel reduce allows to drop unnecessary copying:

```
struct Body {
    std::vector<std::set<int>>& sets;
    std::set<int> result;
    
    void operator()(const blocked_range<size_t>& range) {
        for (size_t i = range.begin(); i != range.end(); ++i) {
            result.merge(sets[i]); // only moving nodes
        }
    }
    
    void join(Body& rhs) {
        result.merge(rhs.result); // also only moving nodes
    }
};
```

The main goal of this patch is to allow rvalues optimization for lambda version of reduce.
This patch is NOT intended to allow reduction of move-only objects (such as `std::unique_ptr`s) since we still need to copy the identity value into each parallel_reduce leaf.

Specification change is also needed for this PR:
* Relax `ParallelReduceFunc` named requirement to allow `Value operator()(const Range& range, Value&& value) const`
* Relax `ParallelReduceReduction` named requirement to allow `Value operator()(Value&& x, Value&& y) const`.

This PR also contains simple test that I guess should be extended to better cover all overloads and algorithms.

Fixes # - _issue number(s) if exists_

- [x] - git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests) for details)_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [ ] bug fix - _change that fixes an issue_
- [x] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [ ] added - _required for new features and some bug fixes_
- [ ] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [x] needs to be updated
- [ ] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

### Notify the following users
@BenFrantzDale
@rfschtkt
Feel free to participate in review and discussions for this patch:)

### CODE REVIEW

File: include/oneapi/tbb/parallel_reduce.h
Comment (BenFrantzDale) [inline include/oneapi/tbb/parallel_reduce.h:402]: Would it be desirable to `&&` qualify this member function? As in
```cpp
[[nodiscard]] Value&& result() noexcept && {
    return std::move(my_value);
}
```
? 
That would force you to say `return std::move(body).result();` elsewhere, which may be clearer. (In general `body.result()` isn't obvious that it is extracting the result from `body`.)

File: include/oneapi/tbb/parallel_reduce.h
Comment (BenFrantzDale) [inline include/oneapi/tbb/parallel_reduce.h:518]: I think this `std::move` does nothing since `body.result()` is an rvalue?

File: test/conformance/conformance_parallel_reduce.cpp
Comment (BenFrantzDale) [inline test/conformance/conformance_parallel_reduce.cpp:60]: `//!` explaining what this struct is for?

File: test/conformance/conformance_parallel_reduce.cpp
Comment (BenFrantzDale) [inline test/conformance/conformance_parallel_reduce.cpp:234]: A move iterator over a vector of `int`s doesn't buy much. Consider for this example making
```cpp
struct MoveOnlyInt {
    int x;
    MoveOnlyInt() = default;
    explicit MoveOnlyInt(int x) : x{x} {}
    MoveOnlyInt(const MoveOnlyInt&) = delete;
    MoveOnlyInt(MoveOnlyInt&& other) : x{std::exchange(other.x, 0)} {}
    MoveOnlyInt& operator=(const MoveOnlyInt&) = delete;
    MoveOnlyInt& operator=(MoveOnlyInt&& other) {
        this->x = std::exchange(other.x, 0);
        return *this;
    }
};
```
and use vectors of `MoveOnlyInt` so the move iterators are required?

Comment (BenFrantzDale) [reply test/conformance/conformance_parallel_reduce.cpp:234]: Could also consider a test case using `std::list` or `std::set` that moves the nodes across, like
```cpp
auto join_body = [](std::set<int>&& x, std::set<int>&& y) {
   while (!y.empty()) {
       x.insert(y.extract(y.begin()));
   }
   return std::move(x);
}
```
(That could use `MoveOnlyInt` too.)

**Review by BenFrantzDale — COMMENTED** [review]
Great to see someone looking at this again! A few suggestions.

File: include/oneapi/tbb/parallel_reduce.h
Comment (kboyarinov) [reply include/oneapi/tbb/parallel_reduce.h]: Since this part of the reduce API was not changed at all, it was a mistake to change this concept. This line was removed.
`parallel_reduce_reduction` and `parallel_reduce_combine` concepts were changed instead.

File: include/oneapi/tbb/parallel_reduce.h
Comment (kboyarinov) [reply include/oneapi/tbb/parallel_reduce.h]: My first thought on this question was removing the `&&` qualifier because as an internal structure, the body would be only move-from the result. But now I guess having this qualifier would be more clear from the move semantics perspective. Thanks for catching that.

File: include/oneapi/tbb/parallel_reduce.h
Comment (kboyarinov) [reply include/oneapi/tbb/parallel_reduce.h]: Changed to `std::move(body).result()`, as discussed in the previous thread.

File: include/oneapi/tbb/parallel_reduce.h
Comment (BenFrantzDale) [reply include/oneapi/tbb/parallel_reduce.h]: Yeah. An alternative is to give it a different name so it’s clearer that it moves out without the `std::move`, but just qualifying it like this feels reasonable.

File: test/conformance/conformance_parallel_reduce.cpp
Comment (BenFrantzDale) [reply test/conformance/conformance_parallel_reduce.cpp:234]: It would probably be good to do a test with a type that isn’t default constructible and/or isn’t movable or copyable in a node-based container — basically all the weird corner cases of irregular types.

File: test/conformance/conformance_parallel_reduce.cpp
Comment (kboyarinov) [reply test/conformance/conformance_parallel_reduce.cpp]: Done

File: test/conformance/conformance_parallel_reduce.cpp
Comment (kboyarinov) [reply test/conformance/conformance_parallel_reduce.cpp]: Unfortunately, we still cannot use move-only types for parallel_reduce since we need to copy the identity into each reduction "leaf". And using `std::vector<MoveOnlyInt>` or list would cause compilation errors even if the identity element is empty. 
I have changed the value type into the integer wrapper that is copyable, but the actual copy constructor should never be called.

Regarding adding new test cases, I think it makes no sense to test both "vector of vectors" and "vector of lists" use-cases. Splicing the lists while doing parallel_reduce is more interesting IMHO, so I left only the vector of lists test.

File: test/conformance/conformance_parallel_reduce.cpp
Comment (BenFrantzDale) [inline test/conformance/conformance_parallel_reduce.cpp:78]: Wouldn't it be better to `=delete;` these?

File: test/conformance/conformance_parallel_reduce.cpp
Comment (BenFrantzDale) [inline test/conformance/conformance_parallel_reduce.cpp:306]: Is the trailing return type helpful versus just leave it to `auto` (or `decltype(auto)`)?

**Review by BenFrantzDale — COMMENTED** [review]
Looks good! I'll leave it to tbb maintainers to approve.

File: test/conformance/conformance_parallel_reduce.cpp
Comment (kboyarinov) [reply test/conformance/conformance_parallel_reduce.cpp:306]: We are forced to use trailing return type, because both `auto` with automatic deduction and `decltype(auto)` are C++14 features and we still need to support C++11.

File: test/conformance/conformance_parallel_reduce.cpp
Comment (kboyarinov) [reply test/conformance/conformance_parallel_reduce.cpp]: I have investigated one more time, and since we don't use `std::vector` directly, only using the wrapper, we can control that actual copy constructor would not be called by the container. So I removed `NeverCopyWrapper` and changed it to be `MoveOnlyWrapper`

File: test/conformance/conformance_parallel_reduce.cpp
Comment (BenFrantzDale) [reply test/conformance/conformance_parallel_reduce.cpp:298]: Gotcha. That's annoying.

**Review by pavelkumbrasev — APPROVED** [review]

**Comment by pavelkumbrasev** [comment]
LGTM!

**Pull request merged**
